lib/require_all.rb

```diff
@@ -92,51 +93,15 @@ def require_all(*args)
       return true
     end
 
-    files.map! { |file_| File.expand_path file_ }
-    files.sort!
-
-    begin
-      failed = []
-      first_name_error = nil
-
-      # Attempt to load each file, rescuing which ones raise NameError for
-      # undefined constants.  Keep trying to successively reload files that 
-      # previously caused NameErrors until they've all been loaded or no new
-      # files can be loaded, indicating unresolvable dependencies.
-      files.each do |file_|
-        begin
-          __require(options[:method], file_)
-        rescue NameError => ex
-          failed << file_
-          first_name_error ||= ex
-        rescue ArgumentError => ex
-          # Work around ActiveSuport freaking out... *sigh*
-          #
-          # ActiveSupport sometimes throws these exceptions and I really
-          # have no idea why.  Code loading will work successfully if these
-          # exceptions are swallowed, although I've run into strange 
-          # nondeterministic behaviors with constants mysteriously vanishing.
-          # I've gone spelunking through dependencies.rb looking for what 
-          # exactly is going on, but all I ended up doing was making my eyes 
-          # bleed.
-          #
-          # FIXME: If you can understand ActiveSupport's dependencies.rb 
-          # better than I do I would *love* to find a better solution
-          raise unless ex.message["is not missing constant"]
-
-          STDERR.puts "Warning: require_all swallowed ActiveSupport 'is not missing constant' error"
-          STDERR.puts ex.backtrace[0..9]
-        end
-      end
-
-      # If this pass didn't resolve any NameErrors, we've hit an unresolvable
-      # dependency, so raise one of the exceptions we encountered.
-      if failed.size == files.size
-        raise first_name_error
-      else
-        files = failed
+    files.map { |file_| File.expand_path file_ }.sort.each do |file_|
+      begin
+        __require(options[:method], file_)
+      rescue NameError => e
+        # Only wrap NameError exceptions for uninitialized constants
+        raise e unless e.instance_of?(NameError) && e.message.include?('uninitialized constant')
+        raise RequireError, "Could not require #{file_} (#{e}). Please require the necessary files"
```

Comment:
It would be cool if we could print out something like this in here:
```
require_all "lib" # a.rb depends on b.rb

# raises an error with a message similar to:
Could not load lib/a.rb, because it has a dependency load order problems. Please resolve it by specify dependencies before anything else like this:
require_all "lib/b.rb", "lib"
```

However, the method "require_all" needs to be correct depending on the usage (e.g. it might be load_all or autoload_all too).

Reply:
I'm not sure that's the correct way of fixing the issue. If `a.rb` depends on `b.rb`, then (IMO) it should declare this and explicitly resolve the dependency by requiring b? It should be necessary (and is not desirable, again IMO) to have knowledge about the dependencies of other code externalised outside of the units themselves?

The error message this gives is sufficient to know which file caused the dependency issue, but we can't do much more as we don't know for sure which file contains the constant we need?

Reply:
Okay, makes sense. Let's keep it as is.